book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,262 @@
+different fork than the one this validiator last rooted. Thus it's critical for
+consistency in `replay_stage` that the order of events when setting a new root
+is:
+
+    1) Write the root to tower
+    
+    2) Write the root to blocktree
+    
+    3) Generate snapshot for that root
+
+3) On startup, the validator checks that the current root in tower exists in
+blocktree. If not (there was a crash between 2-1 and 2-2), then rewrite the
+root to blocktree.
+
+4) On startup, the validator boots from the snapshot `S`, then replays all
+descendant blocks of `S` that exist in this validator's ledger, building banks
+which are then stored in an output `BankForks`. This is done in
+`blocktree_processor.rs`. The root of this `BankForks` is set to `S`.
+
+5) On startup, the validator calculates the first slot `S_n` from which it is
+not locked out for voting. Every validator persists its tower state and must
+consult this state in order to boot safely and resume from a snapshot without
+being slashed. From this tower state and the ancestry information embedded in
+the snapshot, a validiator can derive which banks, are "safe" (See the
+`Determining Vote Safety From a Snapshot and Tower` section for more detals) to
+vote for.
+
+6) Periodically send canary transactions to the cluster using the validator's
+local recent blockhash.
+
+7) Wait for the following criteria. While waiting, set a new root every time
+2/3 of the cluster's stake roots a bank. This allows the validator to prune its
+state and also calculate leader schedules as it moves across epochs.
+
+   1) Verify threshold commitment from the trusted validators on `S`. See the
+   `Trusted Validator Set` section for details.
+
+   2) Observe a canary transaction in a bank that some threshold of trusted
+   validator stake has voted on at least once. Call this trusted bank `T`.
+   
+   3) The current working bank has slot number `S_current` > `S_n`
+
+8) Start the normal voting process, voting for any slot that satisfies
+`S_current` > `S_n`
+
+
+## Trusted Validator Set
+
+The booting validator needs to have a set of trusted validators that it uses to
+validate the snapshot and that it is caught up.  To guarantee that the booting
+validator connects to the cluster, the set needs to contain some threshold of
+validators that are still participating in consensus. This excludes validators
+that are running but are experiencing some attack (such as an eclipse attack)
+that prevents them from receiving slots, voting, or broadcasting their votes.
+
+### Verifying the Cluster is Building from the Snapshot
+
+The booting validator can confirm that the snapshot is valid and that the
+cluster it intends to join is building off of `S` by verifying the signatures
+of votes for `S`. The validator can calculate the commitment of its trusted
+validators, and at some threshold, such as 2/3 of trusted validators are locked
+out at least 2^31 slots, accept the snapshot.
```

Comment:
Let's use `MAX_LOCKOUT - 1` here for clarity as MAX_LOCKOUT is subject to change